src/managers/service.py

```diff
@@ -0,0 +1,191 @@
+"""Utility module containing logic of creation and deletion of managed Kyuubi K8s service."""
+
+import enum
+import functools
+import socket
+import typing
+
+import lightkube
+
+from constants import JDBC_PORT
+from utils.logging import WithLogging
+
+
+class _ServiceType(enum.Enum):
+    """Supported K8s service types."""
+
+    CLUSTER_IP = "ClusterIP"
+    NODE_PORT = "NodePort"
+    LOAD_BALANCER = "LoadBalancer"
+
+
+class ServiceManager(WithLogging):
+    """Utility class containing logic of creation and deletion of managed Kyuubi K8s service."""
+
+    def __init__(self, namespace: str, unit_name: str, app_name: str):
+        self.namespace = namespace
+        self.unit_name = unit_name
+        self.app_name = app_name
+        self.service_name = f"{self.app_name}-service"
+        self.lightkube_agent = lightkube.Client()
+
+    @property
+    def model_service_domain(self) -> str:
+        """K8s service domain for Juju model."""
+        # Example: "kyuubi-k8s-0.kyuubi-k8s-endpoints.my-model.svc.cluster.local"
+        fqdn = socket.getfqdn()
+        # Example: "kyuubi-k8s-0.kyuubi-k8s-endpoints."
+        prefix = f"{self.unit_name.replace('/', '-')}.{self.app_name}-endpoints."
+        assert fqdn.startswith(f"{prefix}{self.namespace}.")
+        # Example: my-model.svc.cluster.local
+        return fqdn.removeprefix(prefix)
+
+    @property
+    def _host(self) -> str:
+        """K8s service hostname for Kyuubi."""
+        # Example: kyuubi-k8s-service.my-model.svc.cluster.local
+        return f"{self.service_name}.{self.model_service_domain}"
+
+    def _get_node_host(self) -> set[str]:
+        """Return the node ports of nodes where units of this app are scheduled."""
+
+        def _get_node_address(node) -> str:
+            # OpenStack will return an internal hostname, not externally accessible
+            # Preference: ExternalIP > InternalIP > Hostname
+            for typ in ["ExternalIP", "InternalIP", "Hostname"]:
+                for address in node.status.addresses:
+                    if address.type == typ:
+                        return address.address
+
+        node = self.get_node(self.unit_name)
+        host = _get_node_address(node)
+        return host
+
+    @functools.cache
+    def get_node(self, unit_name: str) -> lightkube.resources.core_v1.Node:
+        """Return the node for the provided unit name."""
+        node_name = self.get_pod(unit_name).spec.nodeName
+        return self.lightkube_agent.get(
+            res=lightkube.resources.core_v1.Node,
+            name=node_name,
+            namespace=self.namespace,
+        )
+
+    def get_service(self) -> typing.Optional[lightkube.resources.core_v1.Service]:
```

Comment:
**nitpick**: We could tidy up the types a little in this module

```suggestion
    def get_service(self) -> Service | None:
```
or 
```suggestion
    def get_service(self) -> K8sService | None:
```
if we do `from lightkube.resources.core_v1 import Service as K8sService`

The syntax `| None` is not py3.8 compatible, but we already use it elsewhere. We just might increase the py version in `pyproject.toml` instead

Reply:
I'll fix this on upcoming PRs since I don't want the Python version bump to cause unforeseen problems.